group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,1009 @@
+public class CoordinatorRuntime<S extends Coordinator<U>, U> {
+
+    /**
+     * Builder to create a CoordinatorRuntime.
+     *
+     * @param <S> The type of the state machine.
+     * @param <U> The type of the record.
+     */
+    public static class Builder<S extends Coordinator<U>, U> {
+        private LogContext logContext;
+        private CoordinatorEventProcessor eventProcessor;
+        private PartitionWriter<U> partitionWriter;
+        private CoordinatorLoader<U> loader;
+        private CoordinatorBuilderSupplier<S, U> coordinatorBuilderSupplier;
+
+        public Builder<S, U> withLogContext(LogContext logContext) {
+            this.logContext = logContext;
+            return this;
+        }
+
+        public Builder<S, U> withEventProcessor(CoordinatorEventProcessor eventProcessor) {
+            this.eventProcessor = eventProcessor;
+            return this;
+        }
+
+        public Builder<S, U> withPartitionWriter(PartitionWriter<U> partitionWriter) {
+            this.partitionWriter = partitionWriter;
+            return this;
+        }
+
+        public Builder<S, U> withLoader(CoordinatorLoader<U> loader) {
+            this.loader = loader;
+            return this;
+        }
+
+        public Builder<S, U> withCoordinatorStateMachineSupplier(CoordinatorBuilderSupplier<S, U> coordinatorBuilderSupplier) {
+            this.coordinatorBuilderSupplier = coordinatorBuilderSupplier;
+            return this;
+        }
+
+        public CoordinatorRuntime<S, U> build() {
+            if (logContext == null)
+                logContext = new LogContext();
+            if (eventProcessor == null)
+                throw new IllegalArgumentException("Event processor must be set.");
+            if (partitionWriter == null)
+                throw new IllegalArgumentException("Partition write must be set.");
+            if (loader == null)
+                throw new IllegalArgumentException("Loader must be set.");
+            if (coordinatorBuilderSupplier == null)
+                throw new IllegalArgumentException("State machine supplier must be set.");
+
+            return new CoordinatorRuntime<>(
+                logContext,
+                eventProcessor,
+                partitionWriter,
+                loader,
+                coordinatorBuilderSupplier
+            );
+        }
+    }
+
+    /**
+     * The various state that a coordinator for a partition can be in.
+     */
+    enum CoordinatorState {
+        /**
+         * Initial state when a coordinator is created.
+         */
+        INITIAL {
+            @Override
+            boolean canTransitionFrom(CoordinatorState state) {
+                return false;
+            }
+        },
+
+        /**
+         * The coordinator is being loaded.
+         */
+        LOADING {
+            @Override
+            boolean canTransitionFrom(CoordinatorState state) {
+                return state == INITIAL || state == CLOSED || state == FAILED;
```

Comment:
so we can go from closed/failed back to loading?

Reply:
very good question. it actually depends on whether we want to recycle the context or not. after thinking a bit more about it, i think that it is preferable to use reuse it (cause of the snapshot registry). i have updated the state machine and the code to reflect this.